### PR TITLE
osdc: fix mutex assert for !debug builds

### PR DESCRIPTION
**Pull request by batrick**
Fixes: 55652f0819761d410bddcf3688b1c3e10ed64f5b

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/osdc/Journaler.h [hunk 1]
```diff
@@ -529,76 +529,76 @@ class Journaler {
   // ===================
 
   Header get_last_committed() const {
-    ceph_assert(!ceph_mutex_is_locked_by_me(lock));
+    ceph_assert(!ceph::mutex_debugging || !ceph_mutex_is_locked_by_me(lock));
```
Comment (idryomov) [inline src/osdc/Journaler.h:532]: Are these asserts actually useful/worth it?  In all cases they are followed by `lock_guard l(lock);` which if the mutex is already held by the same thread would throw `std::system_error` or just deadlock (as implementations are only encouraged to detect invalid usage).  Either way, it's not hard to diagnose.

There is only one existing instance of such an assert (in `src/common/config_proxy.h`), so it's almost unprecedented too.

Comment (batrick) [reply src/osdc/Journaler.h:532]: > Are these asserts actually useful/worth it? In all cases they are followed by `lock_guard l(lock);` which if the mutex is already held by the same thread would throw `std::system_error` or just deadlock (as implementations are only encouraged to detect invalid usage). Either way, it's not hard to diagnose.

I assume you're citing from: https://en.cppreference.com/w/cpp/thread/mutex/lock

That document says the behavior is undefined so I would say the assert is warranted.

> There is only one existing instance of such an assert (in `src/common/config_proxy.h`), so it's almost unprecedented too.

Yes, I added that too. It has been very helpful in my experience for debugging.

Comment (cbodley) [reply src/osdc/Journaler.h:532]: > I assume you're citing from: https://en.cppreference.com/w/cpp/thread/mutex/lock

does [ceph::mutex_debug::lock()](https://github.com/ceph/ceph/blob/main/src/common/mutex_debug.h#L181) not do this checking when mutex debugging is enabled?

Comment (batrick) [reply src/osdc/Journaler.h:532]: Possibly, but let's not delay getting this merged if it fixes the problem.

**Review by cbodley — APPROVED** [review]

File: src/osdc/Journaler.h (hunk 1 above)
Comment (batrick) [reply src/osdc/Journaler.h:532]: https://github.com/ceph/ceph/pull/60037

**Pull request merged**